File: NinePSharp.Messages.FSharp/messages/Tlopen.fs
namespace NinePSharp.Messages.FSharp

open System
open System.Buffers.Binary
open NinePSharp.Constants
open NinePSharp.Interfaces
open NinePSharp.Messages
open NinePSharp.Messages.FSharp.Protocol

type Tlopen =
    { Size: uint32
      Tag: uint16
      Fid: uint32
      Flags: uint32 }

    interface ISerializable with
        member this.Size = this.Size
        member this.Type = MessageTypes.Tlopen
        member this.Tag = this.Tag
        member this.WriteTo(span: Span<byte>) =
            let mutable offset = 0
            writeHeaders this.Size this.Tag MessageTypes.Tlopen span &offset
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Fid); offset <- offset + 4
            BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), this.Flags)

    static member Parse (data: ReadOnlySpan<byte>) =
        let mutable offset = 0
        let size, tag = readHeaders data &offset
        let fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        offset <- offset + 4
        let flags = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(offset, 4))
        { Size = size; Tag = tag; Fid = fid; Flags = flags }

File: NinePSharp/messages/Tlopen.cs
namespace NinePSharp.Messages;

using System;
using System.Buffers.Binary;
using NinePSharp.Constants;
using NinePSharp.Interfaces;
using NinePSharp.Protocol;

public readonly struct Tlopen : ISerializable
{
    public uint Size { get; }
    public MessageTypes Type => MessageTypes.Tlopen;
    public ushort Tag { get; }
    public uint Fid { get; }
    public uint Flags { get; }

    public Tlopen(uint size, ushort tag, uint fid, uint flags)
    {
        Size = size;
        Tag = tag;
        Fid = fid;
        Flags = flags;
    }

    public Tlopen(ReadOnlySpan<byte> data)
    {
        Size = BinaryPrimitives.ReadUInt32LittleEndian(data[..4]);
        Tag = BinaryPrimitives.ReadUInt16LittleEndian(data.Slice(5, 2));
        Fid = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(NinePConstants.HeaderSize, 4));
        Flags = BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(NinePConstants.HeaderSize + 4, 4));
    }

    public void WriteTo(Span<byte> span)
    {
        span.WriteHeaders(Size, Tag, MessageTypes.Tlopen);
        int offset = NinePConstants.HeaderSize;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Fid); offset += 4;
        BinaryPrimitives.WriteUInt32LittleEndian(span.Slice(offset, 4), Flags);
    }
}
